### PR TITLE
[ICDS Dashboard] Automated Monitors: AWC launched check for Performance Report:

### PR DESCRIPTION
**Pull request by challabeehyv**
- Added validation check for is_launched in AwwIncentive report vs AggAwc
- Added method to send email with csv incase of mismatch

##### SUMMARY
Added awcs is_lanched check after awc aggregation. If fails, sends  a mail with csv containing mismatches

### CODE REVIEW

File: custom/icds_reports/tasks.py [hunk 1]
```diff
@@ -1199,6 +1199,11 @@ def build_incentive_report(agg_date=None):
         agg_date = current_month - relativedelta(months=1)
     for state in state_ids:
         AWWIncentiveReport.aggregate(state, agg_date)
+
+    aggregate_queryset = AWWIncentiveReport.objects.all()
```
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: It would be clearer if all of this data access happened in the check itself. It should not be needed to call this here and pass it in.

File: custom/icds_reports/tasks.py [hunk 2]
```diff
@@ -1209,6 +1214,40 @@ def build_incentive_report(agg_date=None):
                 build_incentive_files.delay(location, agg_date, file_format, 3, location.parent.parent, location.parent)
 
 
+def is_launched_awc_incentive_report(launched_awc_ids):
+    awc_ids = launched_awc_ids.values_list('awc_id', flat=True)
```
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: You modify this queryset a lot, without doing data access (or at least when the access is not necessary). Once this is changed to not pass in the queryset, i think this can be simplified as `AWWIncentiveReport.objects.values_list('awc_id', 'is_launched')` instead of the incremental building done above. As the code is currently written, it looks like we dont even fetch the launch status which is the column that we care about.

File: custom/icds_reports/tasks.py [hunk 3]
```diff
@@ -1209,6 +1214,40 @@ def build_incentive_report(agg_date=None):
...
+    awc_launched_count_from_aggregate = AggAwc.objects.filter(aggregation_level=5, awc_id__in=awc_ids).all()
```
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: These two tables should contain identical lists of awcs, so the filter is not needed and matching a list of 600000 will make the query significantly slower. Also the `.all()` is unnecessary with the django orm. You probably only want to pull awc_id and is_launched here as well (the two columns used in the actual data access below). Same comment as above about incremental building, it is hard to follow

File: custom/icds_reports/tasks.py [hunk 4]
```diff
@@ -1209,6 +1214,40 @@ def build_incentive_report(agg_date=None):
...
+    csv_columns = ['awc_id', 'AwwIncentiveReport', 'AggAwc']
+    csv_data = []
+    for awc in launched_awc_ids:
+        awc_from_aggregate = awc_launched_count_from_aggregate.filter(awc_id=awc[0]).values_list('awc_id',
```
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: This will do one db query per awc (currently around 500000). You should be able to do this as one query outside the loop, and then loop over the returned set. If both queries `AggAwc` and `AWWIncentive` order the returned values by awc_id, then you can loop over them at the same time comparing each entry as you get there.

File: custom/icds_reports/tasks.py [hunk 5]
```diff
@@ -1209,6 +1214,40 @@ def build_incentive_report(agg_date=None):
...
+                                                                                                 'is_launched')[0]
+        is_launched_from_awc = (lambda: False, lambda: True)[awc_from_aggregate[1].lower() == 'yes']()
```
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: This line is difficult to parse, and appears to have the same functionality as just the inner part `awc_from_aggregate[1].lower() == 'yes'`. It would be much easier to read if it was reduced to that.

**Review by calellowitz — CHANGES_REQUESTED** [review]
I think the general idea of this function is correct, but there are a number of improvements to both the readability and performance that should be made before this is merged.

File: custom/icds_reports/tasks.py (hunk 1 above)
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: there are other checks like home conduct and eligibility will use the same queryset. so i thought of creating it here and passing it as a variable. This has been refactored in https://github.com/dimagi/commcare-hq/pull/25366

File: custom/icds_reports/tasks.py (hunk 2 above)
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: This has been refacroed in https://github.com/dimagi/commcare-hq/pull/25366 as more checks are dependent on it
